### "VI. Bioinformatics in R (presentation)"
### author: "Center for Health Data Science, University of Copenhagen"
### source("~/Desktop/HeaDS/GeneExpressDat/Presentation6Functions.R")

### See the files presentation6.html and presentation6.pdf for further explanations. 
### These files are generated by the Rmarkdowm file presentation6.Rmd 




#############

# Gene Expression Analysis in R with DEseq2

#############


## PART 1 - Preliminary Analysis


### Bioconductor open source and open software development

# Bioconductor <https://www.bioconductor.org/> provides tools for computational biology and bioinformatics analysis in R. 
# When we install R-packages we often `install.packages('name_of_package')`, referring to the CRAN <repository(https://cran.r-project.org/> of packages.
# However, sometimes we want a package from `Bioconductor` instead. For this we use the command `BiocManager::install('name_of_package')`. 
# In order to use this installer, we have to download the R-package `BiocManager` e.g. `install.packages('BiocManager')`.  



### Importing the Dataset

# The dataset used for this presentation was acquired from the following github tutorial on RNAseq analysis: <https://combine-australia.github.io/RNAseq-R/06-rnaseq-day1.html>.
# RNA sequencing data generated from luminal and basal cell sub-populations in the mammary gland of three groups of mice:   
    # Control   
    # Pregnant   
    # Lactating

# The objective of the original study (https://pubmed.ncbi.nlm.nih.gov/25730472/) was to identify genes specifically expressed in lactating mammary glands, the gene expression profiles of luminal and basal cells from different developmental stages were compared.


# Set working directory:
setwd("~/Desktop/FromExceltoR/Presentations")


#############


### Load R-packages:
# Data Wrangling
# install.packages("tidyverse")
# install.packages("readxl")
library(tidyverse)
library(readxl)


# For Plotting
# install.packages("ggplot2")
# install.packages("viridis")
library(ggplot2)
library(viridis)


# For DEA
# install.packages("BiocManager")
# BiocManager::install("DESeq2")
library(BiocManager)
library(DESeq2)


# For Enrichment Analysis
# BiocManager::install("clusterProfiler")
# BiocManager::install("org.Mm.eg.db")

library(clusterProfiler)
library(org.Mm.eg.db)


#############


### Reading in data
exprDat <- read_excel("MouseRNAseq.xlsx")
exprInfo <- read_excel("MouseSampleInfo.xlsx")

# Look at the data:
head(exprDat, n=5)
dim(exprDat)

exprInfo


# Convert character columns to factor types:
exprInfo <- exprInfo %>%
  mutate(CellType = as.factor(CellType),
         Status = factor(Status, levels = c("control", "pregnant", "lactate")),
         Status.Type = as.factor(Status.Type))

exprInfo


## Extracting Gene Dataset



#############

### Initial Data Check

# All together:

expr1 <- exprDat %>% 
  dplyr::select(-EntrezGeneID, -GeneName) %>%
  gather() %>% 
  dplyr::select(value) %>%
  mutate(valuelog2 = log2(value+1))

head(expr1, n=5)


# Plot it with ggplot2:

p1 <- ggplot(expr1, aes(value)) + 
  geom_histogram(color="black", fill="grey80", bins=100) + 
  theme_minimal()
p1


p2 <- ggplot(expr1, aes(valuelog2)) + 
  geom_histogram(color="black", fill="grey80", bins=100) + 
  theme_minimal()

p2


# The first plot tells us that we have a lot of 0 counts. Let's try to sample n random genes and plot their count distribution:
expr10 <- exprDat %>%
  dplyr::select(-EntrezGeneID, -GeneName) %>%
  sample_n(.,12) %>% 
  t() %>%
  as_tibble() %>% 
  rename_at(vars(names(.)), ~paste0("Gene", seq(1:12))) %>% 
  gather()

expr10

# Plot n randomly sampled genes:
ggplot(expr10, aes(value)) + 
  geom_histogram(color="black", fill="grey80", bins=20) + 
  theme_minimal() +
  facet_wrap(~key)


# --------------


### Filtering


# Count number of 0s across samples. 
# Filter samples where at least four samples has a count great than 0:

exprDat <- exprDat %>% 
  mutate(nzeros = rowSums(dplyr::select(.,-EntrezGeneID, -GeneName)==0)) %>%
  filter(nzeros <= 8) %>%
  dplyr::select(-nzeros)

# How many genes do we have left:
dim(exprDat)






#############


## PART 2 - Differential Expression Analysis- DESeq2



# Pull out GeneNames and EntrezGeneID for later use.
GeneNames <- exprDat %>% 
  dplyr::select(EntrezGeneID, GeneName)


# Convert to exprDat to a dataframe and make GenNames column into rownames:
exprDat <- exprDat %>%
  dplyr::select(-EntrezGeneID) %>%
  column_to_rownames(., var = "GeneName")


# Make a DESeq2 object:
exprObj <- DESeqDataSetFromMatrix(countData = exprDat,
                              colData = exprInfo,
                              design= ~CellType+Status)

exprObj
# --------------


# Estimating dispersion, gene-wise and mean-dispersion, fitting model and testing:
exprObj <- DESeq(exprObj)

# --------------

# Counts:
head(assay(exprObj))

# Let's have a look at the library sizes and scaling factors:
colSums(assay(exprObj))
#colData(exprObj)


# The count distributions may be dominated by a few genes with very large counts. These genes will drive plotting e.g. heatmaps, PCA analysis etc.
# boxplot(assay(exprObj))
boxplot(assay(exprObj))
boxplot(log2(assay(exprObj)+1))


# Variance stabilizing transformation
# vst returns log2 counts adjusted for sequencing depths. 
# vst removes the dependence of the variance on the mean, particularly the high variance of the log counts when the mean is low.

exprObjvst <- vst(exprObj,blind=TRUE)
boxplot(assay(exprObjvst), xlab="", ylab="Log2 counts per million reads mapped ",las=2)



# --------------

## PCA Plotting:

plotPCA(exprObjvst,intgroup=c("Status"))
plotPCA(exprObjvst,intgroup=c("CellType"))
#plotPCA(exprObjvst,intgroup=c("TypeStatus"))



# --------------


# Test for DE genes between the three groups of mice, adjusted for cell type:

# (I) Lactating vs Control mice
resLC <- results(exprObj, contrast = c("Status", "lactate", "control"), independentFiltering = FALSE)
head(resLC)
dim(resLC)

# Summary and plot of DE analysis results:
summary(resLC)
DESeq2::plotMA(resLC)


# --------------


# Custom function to filter results of DEA. 
# We make a function to save writing the same code three times in a row, one time for each comparison.

# SIGNIFICANT DE GENES:
# Takes as arguments:
# my.res = a dataframe of results from the DEseq results()function
# my.LFC = log fold change cutoff, default is 1.0
# my.cof = adjusted p-value cutoff, default is 0.01 

SigDE <- function(my.res, my.LFC=1.0, my.cof=0.01) {
  my.res <- as.data.frame(my.res) %>%
    rownames_to_column(., var = "GeneName") %>%
    as_tibble() %>%
    left_join(., GeneNames) %>%
    mutate(dir = ifelse(log2FoldChange >= 0, 'up', 'down')) %>%
    filter((log2FoldChange >= my.LFC | log2FoldChange <= -my.LFC) & padj <= my.cof) %>%
    arrange(padj, desc(abs(log2FoldChange)))
  return(my.res)
}

# --------------


# Filter DEA results from comparison of lactating vs control mice using custom function:
resLC <- SigDE(resLC)

# Number of DE genes:
dim(resLC)

# Give it a look:
resLC

# --------------


# (II) Pregnant vs Control mice
resPC <- results(exprObj, contrast = c("Status", "pregnant", "control"),  independentFiltering = FALSE)
#summary(resPC)
#DESeq2::plotMA(resPC)

# Filtering for significant results:
resPC <- SigDE(resPC)

# Number of DE genes:
dim(resPC)


# --------------

# (III) Lactating vs Pregnant mice
resLP <- results(exprObj, contrast = c("Status", "lactate", "pregnant"),  independentFiltering = FALSE)
#summary(resLP)
#DESeq2::plotMA(resLP)

# Filtering for significant results:
resLP <- SigDE(resLP)

# Number of DE genes:
dim(resLP)



#############

## Heatmap Visualization:

# To visually inspect if DE genes identified in our DESeq2 analysis successfully separate the three groups of mice (control, pregnant and lactating), we will make a heatmap. 
# For this we use packages stats and viridis.
# It will not make sense to include all DE genes in this heatmap (3000 genes). Instead pick the top 50 most sig. DE genes based on adj. p-value and logFC.


# Make a vector of unique EntrezGeneIDs (top50):

topDE <- bind_rows(resPC[1:50,], resLC[1:50,], resLP[1:50,]) %>%
  pull(GeneName) %>%
  unique()

length(topDE)

# --------------  

# The expression counts themselves (not logFC) are needed for the heatmap. 
# We use the topDE vector to extract these from the vst normalized DESeq2 object.

# Let's give the data a look:
head(assay(exprObjvst), n=5)


# Extract vst counts for the top 150 (124 unique) genes we want for the heatmap:
resVST <- assay(exprObjvst) %>% 
  as.data.frame() %>%
  rownames_to_column(var = "GeneName")  %>%
  as_tibble() %>%
  filter(GeneName %in% topDE)


# --------------  
# The heatmap function in base R wants gene expression data as a matrix (a dataframe with numeric values only). 

# We extract the GeneNames:
HPnames <- resVST %>% 
  pull(GeneName)

# Convert the tibble into a matrix:
HPdat <- resVST %>%
  dplyr::select(-GeneName) %>%
  as.matrix()

# --------------  
  
# Use the heatmap function to generate a heatmap. 
# We can modify the look of the heatmap as desired, e.g. add column colors, row labels, change color scheme etc.

heatmap(HPdat, labRow=FALSE,
        ColSideColors=exprInfo$CellType.colors, 
        labCol=exprInfo$Status, 
        cexCol=1.2, 
        cexRow = 1.3)


#############


### GO & Pathways Enrichment Analysis

# http://yulab-smu.top/clusterProfiler-book/index.html
# Check that we have pathway information for our species of interest:
search_kegg_organism('mmu', by='kegg_code')



resLCpw <- enrichKEGG(gene = resLC$EntrezGeneID,
                      organism = 'mmu', 
                      universe=unique(GeneNames$EntrezGeneID))
head(resLCpw)


resLCgo <- enrichGO(gene = resLC$EntrezGeneID, 
                    OrgDb = org.Mm.eg.db, 
                    ont = "MF", 
                    pAdjustMethod = "BH",
                    pvalueCutoff  = 0.05,
                    qvalueCutoff  = 0.05, 
                    readable = TRUE,
                    universe=unique(GeneNames$EntrezGeneID))

head(resLCgo)



# --------------

#resPCpw <- enrichKEGG(gene = resPC$EntrezGeneID,organism = 'mmu', universe=unique(GeneNames$EntrezGeneID))
#head(resPCpw)


#resPCgo <- enrichGO(gene = resPC$EntrezGeneID, OrgDb = org.Mm.eg.db, ont = "MF", pAdjustMethod = "BH", pvalueCutoff  = 0.05,qvalueCutoff  = 0.05, readable = TRUE, universe=unique(GeneNames$EntrezGeneID))
#head(resPCgo)


# --------------

#resLPpw <- enrichKEGG(gene = resLP$EntrezGeneID, organism = 'mmu', universe=unique(GeneNames$EntrezGeneID))
#head(resLPpw)


#resLPgo <- enrichGO(gene = resLP$EntrezGeneID, OrgDb = org.Mm.eg.db, ont = "MF", pAdjustMethod = "BH", pvalueCutoff  = 0.01,qvalueCutoff  = 0.05, readable = TRUE, universe=unique(GeneNames$EntrezGeneID))
#head(resLPgo)


# --------------


#resLBpw <- enrichKEGG(gene = resLB$EntrezGeneID, organism = 'mmu', universe=unique(GeneNames$EntrezGeneID))
#head(resLBpw)


#resLBgo <- enrichGO(gene = resLB$EntrezGeneID, OrgDb = org.Mm.eg.db, ont = "MF", pAdjustMethod = "BH", pvalueCutoff  = 0.01,qvalueCutoff  = 0.05, readable = TRUE, universe=unique(GeneNames$EntrezGeneID))
#head(resLBgo)

# --------------

