### "VI. Bioinformatics in R (presentation)"
### author: "Center for Health Data Science, University of Copenhagen"
### source("~/Desktop/HeaDS/GeneExpressDat/Presentation6Functions.R")

### See the files presentation6.html and presentation6.pdf for further explanations. 
### These files are generated by the Rmarkdowm file presentation6.Rmd 




#############

# Gene Expression Analysis in R with DEseq2

#############


## PART 1 - Preliminary Analysis


### Bioconductor open source and open software development

# Bioconductor <https://www.bioconductor.org/> provides tools for computational biology and bioinformatics analysis in R. 
# When we install R-packages we often `install.packages('name_of_package')`, referring to the CRAN <repository(https://cran.r-project.org/> of packages.
# However, sometimes we want a package from `Bioconductor` instead. For this we use the command `BiocManager::install('name_of_package')`. 
# In order to use this installer, we have to download the R-package `BiocManager` e.g. `install.packages('BiocManager')`.  



### Importing the Dataset

# The dataset used for this presentation was acquired from the following github tutorial on RNAseq analysis: <https://combine-australia.github.io/RNAseq-R/06-rnaseq-day1.html>.
# RNA sequencing data generated from luminal and basal cell sub-populations in the mammary gland of three groups of mice:   
    # Control   
    # Pregnant   
    # Lactating

# The objective of the original study (https://pubmed.ncbi.nlm.nih.gov/25730472/) was to identify genes specifically expressed in lactating mammary glands, the gene expression profiles of luminal and basal cells from different developmental stages were compared.


# Set working directory:
setwd("~/Desktop/FromExceltoR_2021/Presentations")


#############


### Load R-packages:
# Data Wrangling
# install.packages("tidyverse")
# install.packages("readxl")
library(tidyverse)
library(readxl)


# For Plotting
# install.packages("ggplot2")
library(ggplot2)



# For DEA
# install.packages("BiocManager")
# BiocManager::install("DESeq2")
library(DESeq2)


#############


### Reading in data
exprDat <- read_excel("MouseRNAseq.xlsx")
exprInfo <- read_excel("MouseSampleInfo.xlsx")

# Look at the data:
head(exprDat, n=5)
dim(exprDat)


# Look at the samples:
exprInfo


# Convert character columns to factor types:
exprInfo <- exprInfo %>%
  mutate(CellType = as.factor(CellType),
         Status = as.factor(Status))

exprInfo


#############

### Initial Data Check

# Let's try to sample 12 (n) random genes and plot their count distribution:

# Sample 16 random rows (genes)
expr16 <- exprDat %>%
  sample_n(.,16)

expr16

# Extract genename
GeneName <- expr16$GeneName

# Gather counts
expr16 <- expr16 %>%
  dplyr::select(-GeneName) %>%
  t() %>% 
  as_tibble() %>% 
  rename_at(vars(names(.)), ~GeneName) %>% 
  gather()



expr16

# Plot n randomly sampled genes:
ggplot(expr16, aes(log2(value+1))) + 
  geom_histogram(color="black", fill="grey80", bins=30) + 
  theme_minimal() +
  facet_wrap(~key)


# --------------


### Filtering


# Count number of 0s across samples. 
# Filter samples where at least four samples has a count great than 0:

#exprDat %>% mutate(nzeros = rowSums(dplyr::select(.,-EntrezGeneID, -GeneName)==0)) %>% dplyr::select(nzeros)

exprDat <- exprDat %>% 
  mutate(nzeros = rowSums(dplyr::select(.,-GeneName)==0)) %>%
  filter(nzeros <= 8) %>%
  dplyr::select(-nzeros)

# How many genes do we have left:
dim(exprDat)






#############


## PART 2 - Differential Expression Analysis- DESeq2

# DEseq2 is developed for handling common issues and biases in expression data. 
# Specifically, differences in sequencing depth and highly variable dispersion of counts between genes.   


# Pull out GeneNames and EntrezGeneID for later use.
GeneNames <- exprDat %>% 
  dplyr::select(GeneName)


# Convert to exprDat to a dataframe and make GenNames column into rownames:
exprDat <- exprDat %>%
  column_to_rownames(., var = "GeneName")

head(exprDat, n=5)



# Make a DESeq2 object:
exprObj <- DESeqDataSetFromMatrix(countData = exprDat,
                              colData = exprInfo,
                              design= ~CellType+Status)

exprObj
# --------------


### Preliminary analysis:

# Counts:
head(assay(exprObj))

# The count distributions may be dominated by a few genes with very large counts. These genes will drive plotting e.g. heatmaps, PCA analysis etc.
# Library sizes:

boxplot(log2(assay(exprObj)+1), las=2)
colSums(assay(exprObj))

# Variance stabilizing transformation
# vst returns log2 counts adjusted for sequencing depths. 

exprObjvst <- vst(exprObj,blind=TRUE)
boxplot(assay(exprObjvst), xlab="", ylab="Log2 counts per million reads mapped ",las=2)
# --------------



## PCA Plotting:

plotPCA(exprObjvst,intgroup="Status")
plotPCA(exprObjvst,intgroup="CellType")
# --------------


### Differential Expression Analysis:


# In brief, DEseq2 fits a generalized linear model (GLM) for each gene in the dataset. 
# DEseq2 adjusts variable gene dispersions by borrowing information across genes to shrink
# gene-wise dispersions towards a common trend to increase accuracy of differential expression testing.

# The GLM fit returns coefficients indicating the overall expression strength of a gene and the log-2 fold change between groups. 


# Estimating dispersion, gene-wise and mean-dispersion, fitting model and testing:
exprObj <- DESeq(exprObj)



# --------------

# Test for DE genes between the three groups of mice, adjusted for cell type:

# (I) Lactating vs Control mice
resLC <- results(exprObj, contrast = c("Status", "lactate", "control"), 
                 independentFiltering = FALSE)

head(resLC)
dim(resLC)

# Summary and plot of DE analysis results:
summary(resLC)
DESeq2::plotMA(resLC)

# --------------



# (II) Pregnant vs Control mice
resPC <- results(exprObj, contrast = c("Status", "pregnant", "control"), 
                 independentFiltering = FALSE)

# Summary and plot of DE analysis results:
#summary(resPC)
#DESeq2::plotMA(resPC)

# --------------


# (III) Lactating vs Pregnant mice
resLP <- results(exprObj, contrast = c("Status", "lactate", "pregnant"),  
                 independentFiltering = FALSE)

# Summary and plot of DE analysis results:
#summary(resLP)
#DESeq2::plotMA(resLP)

# --------------



### Filtering results:

# Bind all DE sets together
# Convert to tibble
# Add columns with GeneNames
# Filter and arrange(order)
resDE <- rbind(resLC, resPC, resLP) %>% 
  as_tibble() %>%
  mutate(GeneName = rep(GeneNames$GeneName, 3)) %>%
  filter((log2FoldChange >= 1.0 | log2FoldChange <= -1.0) & padj <= 0.01) %>%
  arrange(padj, desc(abs(log2FoldChange)))

# --------------

# Number of DE genes:
dim(resDE)
length(unique(resDE$GeneName))


# Give it a look:
resDE

# --------------



#############

## Heatmap Visualization:

# To visually inspect if DE genes identified in our DESeq2 analysis successfully separate the three groups of mice (control, pregnant and lactating), we will make a heatmap. 
# For this we use the heatmap function.
# It will not make sense to include all DE genes in this heatmap (almost 3000 unique genes). Instead pick the top 100 most sig. DE genes based on adj. p-value and logFC.


# Make a vector of unique EntrezGeneIDs (top100):

top100 <- resDE[1:100,] %>%
  pull(GeneName) %>%
  unique()


length(top100)

# --------------  

# The expression counts themselves (not logFC) are needed for the heatmap. 
# We use the topDE vector to extract these from the vst normalized DESeq2 object.

# Let's give the data a look:
head(assay(exprObjvst), n=5)


# Extract vst counts for the top 150 (124 unique) genes we want for the heatmap:
resVST <- assay(exprObjvst) %>% 
  as.data.frame() %>%
  rownames_to_column(var = "GeneName")  %>%
  as_tibble() %>%
  filter(GeneName %in% top100)


# --------------  
# The heatmap function in base R wants gene expression data as a matrix (a dataframe with numeric values only). 

# We extract the GeneNames:
HPnames <- resVST %>% 
  pull(GeneName)

# Convert the tibble into a matrix:
HPdat <- resVST %>%
  dplyr::select(-GeneName) %>%
  as.matrix()

# --------------  
  
# Use the heatmap function to generate a heatmap. 
# We can modify the look of the heatmap as desired, e.g. add column colors, row labels, change color scheme etc.

heatmap(scale(HPdat),
        ColSideColors=exprInfo$CellType.colors, 
        labCol=exprInfo$Status,
        labRow = HPnames,
        cexCol=1.2, 
        cexRow = 1.0)
