### "VI. Bioinformatics in R (presentation)"
### author: "Center for Health Data Science, University of Copenhagen"
### source("~/Desktop/HeaDS/GeneExpressDat/Presentation6Functions.R")

### See the files presentation6.html and presentation6.pdf for further explanations. 
### These files are generated by the Rmarkdowm file presentation6.Rmd 




#############

# Gene Expression Analysis in R with DEseq2

#############


## PART 1 - Preliminary Analysis


### Bioconductor open source and open software development

# Bioconductor <https://www.bioconductor.org/> provides tools for computational biology and bioinformatics analysis in R. 
# When we install R-packages we often `install.packages('name_of_package')`, referring to the CRAN <repository(https://cran.r-project.org/> of packages.
# However, sometimes we want a package from `Bioconductor` instead. For this we use the command `BiocManager::install('name_of_package')`. 
# In order to use this installer, we have to download the R-package `BiocManager` e.g. `install.packages('BiocManager')`.  



### Importing the Dataset

# The dataset used for this presentation was acquired from the following github tutorial on RNAseq analysis: <https://combine-australia.github.io/RNAseq-R/06-rnaseq-day1.html>.
# RNA sequencing data generated from luminal and basal cell sub-populations in the mammary gland of three groups of mice:   
    # Control   
    # Pregnant   
    # Lactating

# The objective of the original study (https://pubmed.ncbi.nlm.nih.gov/25730472/) was to identify genes specifically expressed in lactating mammary glands, the gene expression profiles of luminal and basal cells from different developmental stages were compared.


# Set working directory:
setwd("~/Desktop/FromExceltoR/Presentations")


#############


### Load R-packages:
# Data Wrangling
# install.packages("tidyverse")
# install.packages("readxl")
library(tidyverse)
library(readxl)


# For Plotting
# install.packages("ggplot2")
library(ggplot2)



# For DEA
# install.packages("BiocManager")
# BiocManager::install("DESeq2")
library(DESeq2)


#############


### Reading in data
exprDat <- read_excel("MouseRNAseq.xlsx")
exprInfo <- read_excel("MouseSampleInfo.xlsx")

# Look at the data:
exprDat
dim(exprDat)


# Look at the samples:
exprInfo


# Convert character columns to factor types:
exprInfo <- exprInfo %>%
  mutate(CellType = as.factor(CellType),
         Status = as.factor(Status))

exprInfo


#############

### Initial Data Check

# Let's try to sample 12 (n) random genes and plot their count distribution:

# Sample 16 random rows (genes)
expr16 <- exprDat %>%
  sample_n(.,16)

expr16
dim(expr16)


# Gather counts
expr16 <- expr16 %>%
  column_to_rownames(var = "GeneName") %>%
  t() %>% 
  as_tibble() %>% 
  gather()


expr16

# Plot n randomly sampled genes:
ggplot(expr16, aes(log2(value+1))) + 
  geom_histogram(color="black", fill="grey80", bins=30) + 
  theme_minimal() +
  facet_wrap(~key)


# --------------


### Filtering


# Count number of samples with min. count size of 5 for a given gene. 
# Filter for genes were min. 4 samples have a count equal to or greater than 5.

#exprDat %>% mutate(nzeros = rowSums(dplyr::select(.,-EntrezGeneID, -GeneName)==0)) %>% dplyr::select(nzeros)

exprDat <- exprDat %>% 
  mutate(nzeros = rowSums(dplyr::select(.,-GeneName) >= 4)) %>%
  filter(nzeros <= 8) %>%
  dplyr::select(-nzeros)

# How many genes do we have left:
dim(exprDat)






#############


## PART 2 - Differential Expression Analysis- DESeq2

# DEseq2 is developed for handling common issues and biases in expression data. 
# Specifically, differences in sequencing depth and highly variable dispersion of counts between genes.   


# Pull out GeneNames and EntrezGeneID for later use.
#GeneNames <- exprDat %>% 
#  dplyr::select(GeneName)


# Convert to exprDat to a dataframe and make GenNames column into rownames:
exprDat <- exprDat %>%
  column_to_rownames(., var = "GeneName")

head(exprDat, n=5)



# Make a DESeq2 object:
exprObj <- DESeqDataSetFromMatrix(countData = exprDat,
                              colData = exprInfo,
                              design= ~CellType+Status)

exprObj
# --------------


### Preliminary analysis:

# Counts:
head(assay(exprObj))

# The count distributions may be dominated by a few genes with very large counts. These genes will drive plotting e.g. heatmaps, PCA analysis etc.
# Library sizes:


boxplot(log2(assay(exprObj)+1), las=2)
colSums(assay(exprObj))

# Variance stabilizing transformation
# vst returns log2 counts adjusted for sequencing depths. 

exprObjvst <- vst(exprObj,blind=TRUE)
boxplot(assay(exprObjvst), xlab="", ylab="Log2 counts per million reads mapped ",las=2)
# --------------



## PCA Plotting:

plotPCA(exprObjvst,intgroup="Status")
plotPCA(exprObjvst,intgroup="CellTypes")
# --------------


### Differential Expression Analysis:


# In brief, DEseq2 fits a generalized linear model (GLM) for each gene in the dataset. 
# DEseq2 adjusts variable gene dispersions by borrowing information across genes to shrink
# gene-wise dispersions towards a common trend to increase accuracy of differential expression testing.

# The GLM fit returns coefficients indicating the overall expression strength of a gene and the log-2 fold change between groups. 


# Estimating dispersion, gene-wise and mean-dispersion, fitting model and testing:
exprObj <- DESeq(exprObj)



# --------------

# Test for DE genes between the three groups of mice, adjusted for cell type:

# (I) Lactating vs Control mice
resLC <- results(exprObj, contrast = c("Status", "lactate", "control"), 
                 independentFiltering = FALSE) 

resLC

dim(resLC)
class(resLC)
  
# Summary and plot of DE analysis results:
summary(resLC)
DESeq2::plotMA(resLC)



resLC <-  resLC %>% 
  as.data.frame() %>%
  rownames_to_column(., var = "GeneName")

# --------------



# (II) Pregnant vs Control mice
resPC <- results(exprObj, contrast = c("Status", "pregnant", "control"), 
                 independentFiltering = FALSE) 

# Summary and plot of DE analysis results:
#summary(resPC)
#DESeq2::plotMA(resPC)

resPC <- resPC %>% 
  as.data.frame() %>%
  rownames_to_column(., var = "GeneName")
# --------------


# (III) Lactating vs Pregnant mice
resLP <- results(exprObj, contrast = c("Status", "lactate", "pregnant"),  
                 independentFiltering = FALSE)

# Summary and plot of DE analysis results:
#summary(resLP)
#DESeq2::plotMA(resLP)

resLP <- resLP %>% 
  as.data.frame() %>%
  rownames_to_column(., var = "GeneName")


# --------------



### Filtering results:

# Bind all DE sets together
# Convert to tibble with column GeneNames
# Filter and arrange(order)


resDE <- rbind(resLC, resPC, resLP) %>% 
  as_tibble() %>%
  mutate(pair = c(rep("Lactate.Control", nrow(exprDat)),
                  rep("Pregnant.Control", nrow(exprDat)),
                  rep("Lactate.Pregnant", nrow(exprDat)))) %>%
  filter((log2FoldChange >= 1.0 | log2FoldChange <= -1.0) & padj <= 0.05)

# --------------


# Give it a look:
resDE

# Number of DE genes:
dim(resDE)

resDE %>% 
  pull(GeneName) %>% 
  unique() %>% 
  length()





